Energy API refactor

Technici4n opened the pull request:
Most of the magic now happens in the new [`EnergyStorage`](https://github.com/Technici4n/Energy/tree/transactions/src/main/java/team/reborn/energy/api/EnergyStorage.java). I have chosen that name for consistency with fabric's `Storage<T>`.

I have deprecated for removal but kept the old classes for now to avoid hard breaking existing mods, so there are two `EnergyStorage`s. Fortunately, the older one is deprecated so intellij suggests the newer one first.

TODO:
- [x] Wait for FabricMC/Fabric#1553 to be merged, and then update `build.gradle` and `fabric.mod.json` accordingly.
- [x] Document `SimpleEnergyStorage`.
- [x] Write and document `SimpleEnergyItemStorage`.
- [x] Update README.
- [x] Update tests.
- [x] Port TR to check that the API is usable.

Feel free to question any change... I don't mean to enforce anything. ^^

Technici4n committed:
WIP API refactor

modmuss50 commented on build.gradle:
```diff
@@ -40,41 +26,60 @@ logger.lifecycle("Building Energy: " + version)
-	options.encoding = "UTF-8"
+tasks.withType(JavaCompile).configureEach {
+	it.options.encoding = "UTF-8"
+	it.options.release =16
```
If we are doing this make sure to set the min version in the fabric.mod.json file, along with mc 1.17

Technici4n replied:
Done.

modmuss50 commented on src/main/java/team/reborn/energy/Energy.java:
```diff
@@ -6,6 +6,10 @@
+/**
+ * @deprecated Use the new energy API instead. (Under the team/reborn/energy/api package).
+ */
+@Deprecated(forRemoval = true)
```
This is fine, would be nice to remove them from the source tree, I can possibly do some gradle stuff to shade this in for now, or just move to another source set.

Technici4n replied:
Not sure what the best approach is... I don't want to complicate the build logic too much as I don't really understand it.

modmuss50 reviewed (commented):
Api looks great 👍

Just a few minor house keeping things. Will wait for the fab api before merging. And will wait for a compatible TR before releasing. (Let me know if you want help with this)

Technici4n committed:
Many changes

Add sided base implementation, write item base implementation, and document those.
Add testmod.
Remove old test and jmh benchmark.

Technici4n committed:
Update transfer API to use its recent helper additions

Technici4n committed:
Fix EnergyStorageUtil#move implementation error

Technici4n committed:
Update transfer API

Technici4n committed:
Add EnergyStorageUtil#isEnergyStorage to test if ItemStacks are energy storages

Technici4n committed:
Add SimpleBatteryItem

Technici4n committed:
Javadoc

Technici4n committed:
Update README

Technici4n committed:
Update to released fabric API

Technici4n commented on src/main/java/team/reborn/energy/api/base/SimpleItemEnergyStorage.java:
```diff
@@ -0,0 +1,135 @@
+
+	@Override
+	public long getAmount() {
+		if (hasStartingItem()) {
```
Wait before merging, I'll add something to avoid having to put these checks all over the place.

Technici4n committed:
Add DelegatingEnergyStorage, rework the base item energy storage, add EnergyStorage#EMPTY

Technici4n committed:
Expand DelegatingEnergyStorage
